src/o365/graph/commands/report/report-teamsdeviceusageusercounts.spec.ts

```diff
@@ -0,0 +1,239 @@
+      try {
+        assert(trackEvent.called);
+        done();
+      }
+      catch (e) {
+        done(e);
+      }
+    });
+  });
+
+  it('logs correct telemetry event', (done) => {
+    cmdInstance.action = command.action();
+    cmdInstance.action({ options: {} }, () => {
+      try {
+        assert.equal(telemetry.name, commands.REPORT_TEAMSDEVICEUSAGEUSERCOUNTS);
+        done();
+      }
+      catch (e) {
+        done(e);
+      }
+    });
+  });
+
+  it('aborts when not logged in to Microsoft Graph', (done) => {
+    auth.service = new Service();
+    auth.service.connected = false;
+    cmdInstance.action = command.action();
+    cmdInstance.action({ options: { debug: true } }, (err?: any) => {
+      try {
+        assert.equal(JSON.stringify(err), JSON.stringify(new CommandError('Log in to the Microsoft Graph first')));
+        done();
+      }
+      catch (e) {
+        done(e);
+      }
+    });
+  });
+
+  it('fails validation if period option is not passed', () => {
+    const actual = (command.validate() as CommandValidate)({ options: { } });
+    assert.notEqual(actual, true);
+  });
+
+  it('fails validation on invalid period', () => {
+    const actual = (command.validate() as CommandValidate)({ options: { period: 'abc' } });
+    assert.notEqual(actual, true);
+  });
+
+  it('passes validation on valid \'D7\' period', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        period: 'D7'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation on valid \'D30\' period', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        period: 'D30'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation on valid \'D90\' period', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        period: 'D90'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('passes validation on valid \'180\' period', () => {
+    const actual = (command.validate() as CommandValidate)({
+      options: {
+        period: 'D90'
+      }
+    });
+    assert.equal(actual, true);
+  });
+
+  it('gets the number of Microsoft Teams daily unique users by device type for the given period', (done) => {
+    sinon.stub(request, 'get').callsFake((opts) => {
+      if (opts.url === `https://graph.microsoft.com/v1.0/reports/getTeamsDeviceUsageUserCounts(period='D7')`) {
+        return Promise.resolve('Report Refresh Date,Web,Windows Phone,Android Phone,iOS,Mac,Windows,Report Date,Report Period');
+      }
+
+      return Promise.reject('Invalid request');
+    });
+
+    auth.service = new Service();
+    auth.service.connected = true;
+    auth.service.resource = 'https://graph.microsoft.com';
+    cmdInstance.action = command.action();
+    cmdInstance.action({ options: { debug: false, period: 'D7' } }, () => {
+      try {
+        assert(1 === 1);
```

Comment:
This is not a valid test since it will always pass. Please change it. You can change it to use stub and check if an https request is being executed with the correct url like:

```
  it('gets the number of Microsoft Teams daily unique users by device type for the given period', (done) => {
    // @Velin: define the stub for the requests class
    const requestStub: sinon.SinonStub = sinon.stub(request, 'get').callsFake((opts) => {
      if (opts.url === `https://graph.microsoft.com/v1.0/reports/getTeamsDeviceUsageUserCounts(period='D7')`) {
        return Promise.resolve('Report Refresh Date,Web,Windows Phone,Android Phone,iOS,Mac,Windows,Report Date,Report Period');
      }

      return Promise.reject('Invalid request');
    });

    auth.service = new Service();
    auth.service.connected = true;
    auth.service.resource = 'https://graph.microsoft.com';
    cmdInstance.action = command.action();
    cmdInstance.action({ options: { debug: false, period: 'D7' } }, () => {
      try {
    // @Velin: check if the request has been called with the right url
        assert.equal(requestStub.lastCall.args[0].url, "https://graph.microsoft.com/v1.0/reports/getTeamsDeviceUsageUserDetail(period='D7')");
...
```

Here is more information on stubs and spies:
https://sinonjs.org/releases/latest/stubs/
https://sinonjs.org/releases/latest/spies/